### PR TITLE
Add check-sparse-tensor-invariants flag to Context.

### PR DESCRIPTION
**Pull request by pearu**
This PR adds "check sparse tensor invariants" flag to Context that when enabled will trigger sparse tensor data invariants checks in unsafe methods of constructing sparse COO/CSR/CSC/BSR/BSC tensors. The feature includes the following changes to UI:

- `torch.enable_check_sparse_tensor_invariants` and `torch.is_check_sparse_tensor_invariants_enabled` functions to globally enable/disable the invariant checks and to retrieve the state of the feature, respectively
- `torch.sparse_coo/csr/csc/bsr/bsc/compressed_tensor` functions have a new optional argument `check_invariants` to enable/disable the invariant checks explicitly. When the `check_invariants` argument is specified, the global state of the feature is temporarily overridden. 

The PR also fixes https://github.com/pytorch/pytorch/issues/90833

# Main issue

*The following content is outdated after merging the PRs in this ghstack but kept for the record.*

The importance of this feature is that when enabling the invariants checks by default, say, via

<details>

```
$ git diff
diff --git a/torch/__init__.py b/torch/__init__.py
index c8543057c7..19a91d0482 100644
--- a/torch/__init__.py
+++ b/torch/__init__.py
@@ -1239,3 +1239,8 @@ if 'TORCH_CUDA_SANITIZER' in os.environ:
 
 # Populate magic methods on SymInt and SymFloat
 import torch.fx.experimental.symbolic_shapes
+
+# temporarily enable sparse tensor arguments validation in unsafe
+# constructors:
+
+torch._C._set_check_sparse_tensor_invariants(True)
```

</details>

a massive number of test failures/errors occur in test_sparse_csr.py tests:
```
$ pytest -sv test/test_sparse_csr.py
<snip>
==== 4293 failed, 1557 passed, 237 skipped, 2744 errors in 69.71s (0:01:09) ====
```
that means that we are silently constructing sparse compressed tensors that do not satisfy the sparse tensor invariants. In particular, the following errors are raised:

```
AssertionError: "resize_as_sparse_compressed_tensor_: self and src must have the same layout" does not match "expected values to be a strided and contiguous tensor"

RuntimeError: CUDA error: device-side assert triggered

RuntimeError: `col_indices[..., crow_indices[..., i - 1]:crow_indices[..., i]] for all i = 1, ..., nrows are sorted and distinct along the last dimension values` is not satisfied.

RuntimeError: expected col_indices to be a strided and contiguous tensor

RuntimeError: expected row_indices to be a strided and contiguous tensor

RuntimeError: expected values to be a strided and contiguous tensor

RuntimeError: for_each: failed to synchronize: cudaErrorAssert: device-side assert triggered

RuntimeError: tensor dimensionality must be sum of batch, base, and dense dimensionalities (=0 + 2 + 0) but got 3
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91331
* __->__ #90849

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Comment by pearu** [comment]
Pinging @cpuhrsch @amjames @nikitaved for awarness.

**Comment by nikitaved** [comment]
Thank you, @pearu , this is amazing!

**Comment by amjames** [comment]
With the addition of this flag, could we just drop the safe/unsafe dichotomy altogether? A singular API could provide the ability to be either safe, or fast.

**Comment by pearu** [comment]
> With the addition of this flag, could we just drop the safe/unsafe dichotomy altogether? A singular API could provide the ability to be either safe, or fast.

Yes, I was also thinking about this. There exist two contexts where sparse tensors are constructed:
1. "user context" where a sparse tensor constructor is called by a user program
2. "PyTorch internal context" where a sparse tensor constructor is called in some PyTorch code

In context 1, the invariants checks should be always enabled to catch incorrect user inputs without crashing the Python process.

In context 2, the invariants checks should be enabled when executing PyTorch tests, otherwise, the checks can be disabled for efficiency while trusting on PyTorch tests coverage that should prove/guarantee the correct usage of constructors without invariant checks.

It is not always possible to determine if the execution happens in context 1 or 2, especially on the Python side. The user-facing constructor functions (`torch.sparse_csr_tensor`, etc) always have and should have the invariant checks enabled while PyTorch internal Python code may use the corresponding unsafe methods.

**Comment by amjames** [comment]
I think the same philosophy could be applied to both contexts though. We want to apply the checks in testing to make sure we have written our code that invokes the constructors correctly. Users would want the same thing, enable checks in testing or while debugging but ultimately have them off wherever possible in production.

**Comment by pearu** [comment]
> I think the same philosophy could be applied to both contexts though. We want to apply the checks in testing to make sure we have written our code that invokes the constructors correctly. Users would want the same thing, enable checks in testing or while debugging but ultimately have them off wherever possible in production.

Yes, makes sense. The above implies that the flag should have three states:
1. enable invariant checks in public constructors and no checks in unsafe constructors
2. enable invariant checks in unsafe constructors (recall public constructors call unsafe constructors)
3. disable invariant checks in all circumstances

Btw, one of my points above was that even with this flag we cannot eliminate unsafe constructors.

**Comment by amjames** [comment]
This would be a bit more fine-grained.  In this scenario the unsafe constructors would no longer need to be exported correct?

**Comment by pearu** [comment]
This PR discovered a bug in CSR->BSR conversion: https://github.com/pytorch/pytorch/issues/90910

**Comment by pearu** [comment]
This PR revealed another bug in CSR->CSC conversion when using int32 indices: https://github.com/pytorch/pytorch/issues/91007

**Comment by pearu** [comment]
This PR revealed another issue with requiring full contiguity on batch sparse compressed tensor indices and values: https://github.com/pytorch/pytorch/issues/91062

**Review by amjames — APPROVED** [review]
One nit in-line, but overall looks great! 

If you are up for some extra cleanup. There are a few places where the validation checks are performed manually that should now be possible to remove in favor building the check into the constructor calls with the context manager.

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -2233,6 +2233,22 @@ def setUp(self):
...
+
+        # Enable invariant checks for all sparse tensors constructions
+        # including the unsafe ones. If this is not desired for some
+        # test case, use check_invariants=False optional argument to
+        # sparse tensor constructors or
+        # @CheckSparseTensorInvariants(False) decorator from
+        # test/test_sparse.py to disable the invariant checks.
+        torch.enable_check_sparse_tensor_invariants(True)
+
+    def dearDown(self):
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_utils.py:2248]: nit: tearDown?

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -2233,6 +2233,22 @@ def setUp(self):
...
+        # restored from dearDown:
+        self._check_invariants = torch.is_check_sparse_tensor_invariants_enabled()
+
+        # Enable invariant checks for all sparse tensors constructions
+        # including the unsafe ones. If this is not desired for some
+        # test case, use check_invariants=False optional argument to
+        # sparse tensor constructors or
+        # @CheckSparseTensorInvariants(False) decorator from
+        # test/test_sparse.py to disable the invariant checks.
+        torch.enable_check_sparse_tensor_invariants(True)
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_utils.py:2246]: Technically you could pass `False`. Wouldn't `set_[...]` make more sense than `enable_[...]`?

File: torch/__init__.py [hunk 3]
```diff
@@ -780,6 +782,36 @@ def is_warn_always_enabled():
...
+      both in Python and ATen.
+
+      The flag can be locally overridden by the ``check_invariants``
+      optional argument of the sparse tensor constructor functions.
+
+    Args:
+        enable (:class:`bool`): If True, enable invariants checks,
+                                otherwise, disable the checks.
+    """
+    torch._C._set_check_sparse_tensor_invariants(enable)
```
Comment (cpuhrsch) [inline torch/__init__.py:808]: It's a bit of a nit, but why use `_set` for the internal version and then `enable` for the public?

Also, maybe we can consider only using a Python context manager as the public API?

So then people can do 
```
with torch.sparse.check_sparse_tensor_invariants():
    run_my_model()
```

It's also easier to move a context manager like this to a block of code where they expect an issue or around an op they wrote themselves or to use it as a decorator for a test.

File: test/test_sparse.py [hunk 4]
```diff
@@ -4077,8 +4078,138 @@ def test_basic(self):
         self.assertEqual(r.values(), torch.empty(0, 4, device='meta'))
 
 
+class CheckSparseTensorInvariants(object):
```
[the remark this answers is not in the document]
Comment (cpuhrsch) [reply test/test_sparse.py:4081]: 100% agree that this is of good general utility. I think it's the better public API overall than exposing the getter and setters.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
This is great! It'll serve as an easy to use diagnostic we can advise people by default to try when creating a Github issue and has already spawned lots of fixes all over the place.

Most of my comments are around the public API. I think we should just have a single context manager in the `torch.sparse` namespace for now.

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2248]: Ah, right. Fixed. Thanks!

File: torch/testing/_internal/common_utils.py (hunk 2 above)
Comment (pearu) [reply torch/testing/_internal/common_utils.py:2246]: Yes, it makes sense.

`torch.enable_check_sparse_tensor_invariants(<enabled>)` is now replaced with a static method pair:
```
torch.sparse.check_sparse_tensor_invariants.enable()
torch.sparse.check_sparse_tensor_invariants.disable()
```

File: torch/__init__.py (hunk 3 above)
Comment (pearu) [reply torch/__init__.py:808]: There exists now three ways to enable/disable invariant checks:

1. Using context manager:
   ```python
   with torch.sparse.check_sparse_tensor_invariants(enable=True):
       run_my_model()
   ```
2. Using a procedural approach (when context manager is inconvenient due to the required indent, useful in command prompt, for instance):
   ```python
   prev_is_enabled = torch.sparse.check_sparse_tensor_invariants.is_enabled()
   torch.sparse.check_sparse_tensor_invariants.enable()
   run_my_model()
   if not prev_is_enabled:
       torch.sparse.check_sparse_tensor_invariants.disable()
   ```
3. Via decorator (useful in tests):
   ```python
   @torch.sparse.check_sparse_tensor_invariants(enable=True)
   def run_my_model():
       ...
       
   run_my_model()
   ```

Comment (pearu) [reply torch/__init__.py:808]: And the forth way is to use `check_invariants` kw argument in a sparse tensor constructor call.

**Comment by pearu** [comment]
Thanks, @amjames @cpuhrsch for your feedback. I have updated the PR accordingly: now `torch.sparse.check_sparse_tensor_invariants` together with `check_invariants` kw argument of constructors, provide the public API to manage the invariant checking.

**Review by amjames — APPROVED** [review]
LGTM!

File: torch/sparse/__init__.py [hunk 5]
```diff
@@ -357,3 +358,102 @@ def sum(input: Tensor, dim: DimOrDims = None,
...
+
+class check_sparse_tensor_invariants(object):
+    """A tool to control checking sparse tensor invariants.
+
+The following options exists to manage sparsr tensor invariants
+checking in sparse tensor construction:
+
+1. Using a context manager::
+
+     with torch.sparse.check_sparse_tensor_invariants(enable=True):
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:371]: nit: This makes it look like you have to set `enable=True`, but actually it is set to `True` by default already. It makes the docs look a bit saner if the user doesn't have to repeat their intent with the API call or put differently, we don't have `enable=False` by default.

File: torch/sparse/__init__.py [hunk 6]
```diff
@@ -357,3 +358,102 @@ def sum(input: Tensor, dim: DimOrDims = None,
...
+    torch.sparse.check_sparse_tensor_invariants.enable()
+
+    run_my_model()
+
+    if not prev_checks_enabled:
+        torch.sparse.check_sparse_tensor_invariants.disable()
+
+3. Using function decoration::
+
+    @torch.sparse.check_sparse_tensor_invariants(enable=True)
```
Comment (cpuhrsch) [inline torch/sparse/__init__.py:386]: Same here. I think it makes for a better example if we remove the explicit `enable=True`.

**Review by cpuhrsch — APPROVED** [review]

File: torch/sparse/__init__.py (hunk 5 above)
Comment (pearu) [reply torch/sparse/__init__.py]: Fixed.

File: torch/sparse/__init__.py (hunk 6 above)
Comment (pearu) [reply torch/sparse/__init__.py]: Fixed.

**Comment by cpuhrsch** [comment]
@pearu the error is 
```
stderr: ld.lld: error: undefined symbol: at::native::DispatchStub<void (*)(at::Tensor const&, at::Tensor const&, c10::Scalar const&, c10::Scalar const&, at::Tensor const&), at::native::sampled_addmm_sparse_csr_stub>::DEFAULT

```

This almost seems unrelated? Maybe we just need to rebase and try again.

**Comment by pearu** [comment]
@cpuhrsch could you stamp https://github.com/pytorch/pytorch/pull/92094 that replaces this PR.

**Comment by pearu** [comment]
Closing as its replacement https://github.com/pytorch/pytorch/pull/92094 has landed.

**Pull request closed without merging**